### PR TITLE
Add support for installing with a custom Apache Spark distribution

### PR DESCRIPTION
**Pull request by tmckayus**

### CODE REVIEW

**Comment by tmckayus** [comment]
Is there a way to control the order of the how do I articles? It would be nice of the simple install should up before the "choose my own spark" install.

**Comment by tmckayus** [comment]
I have an idea for a refinement on this, where resources.yaml and resources-is.yaml both work with imagestreams and there is only one template ...

**Comment by tmckayus** [comment]
@elmiko @crobby @wilb ptal

The how_do_i instructions should be enough to show you how to use it, if not let me know! I may work up some CI tests for rad-image as well

**Comment by crobby** [comment]
Building the images works nicely for me
But when I try to run a pyspark (2.7) app, it seem to try to use "openshift-spark:complete" as the image, but pulling it fails.  
The openshift-spark:complete image seems to have been built, but it's 172.30.1.1:5000/<project>/openshift-spark:complete.....whereas the cluster attempt is trying to just use openshift-spark:complete.  Could be a lookup issue (I'm on v3.11  oc cluster up)

**Comment by crobby** [comment]
Ok, it *does* seem like my issue is indeed the infamous local lookup bug with my oc cluster up instance.  Tweaking the config map via:  oc edit configmap default-oshinko-cluster-config   and changing the sparkimage to reference the full image <repo>/<project>/<image>:<tag> allows my jobs to run.

**Comment by tmckayus** [comment]
I have an idea to fix the openshift-spark references in the configmaps, pending

**Comment by elmiko** [comment]
just so i understand clearly, the updates to resources.yaml now point all the template to ImageStreams, which in turn are backed by the community images?

**Comment by tmckayus** [comment]
@elmiko yes the default imagestreams are pointed at the full community images.

When you do an image completion, you end up with a new tag in the imagestream referencing the image you just built.  The "rad-image use" command just mods the template to call out a different tag, but they always reference the imagestream.

This gets us consistent behavior whether you're just using community images out of the box, or you want to roll your own, and you only have 1 template set.

**Comment by elmiko** [comment]
awesome, thanks for the explanation

File: _howdoi/choose-my-spark-distribution.adoc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+. Build images using the Spark distribution you've selected. This will create new `complete` tags on the radanaltyics.io image streams (it may take a while since it will build multiple images). Run `rad-image build -h` for additional information on the build command.
+
+    ./rad-image build https://archive.apache.org/dist/spark/spark-2.3.0/spark-2.3.0-bin-hadoop2.7.tgz
+
+. Make the radanalytics.io templates and configmaps use the `complete` tag
+
+    ./rad-image use complete
+
+. To use the default tags again, do this
```
Comment (elmiko) [inline _howdoi/choose-my-spark-distribution.adoc]: i think we should move this under the item for the `rad-image use` step, just to make it clear that it is an optional step.

File: _howdoi/choose-my-spark-distribution.adoc [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    ./rad-image use -d
+
+. To build images with a tag other than `complete`, specify the `-t TAG` option on the build command
```
Comment (elmiko) [inline _howdoi/choose-my-spark-distribution.adoc]: likewise, i think this step should be moved under the item for the initial build step. just to make it clear these are options.

**Comment by tmckayus** [comment]
@elmiko oshinko-webui will be okay as long as it reads default-oshinko-cluster-config to override the default configuration (which it may not). Alternatively, you scan specify the sparkimage on the cluster launch form so you could select a completed image stream using the pullspec from the project.

**Comment by tmckayus** [comment]
> i was just trying to think about corner cases where we might hit a not "easy button" type interaction. ideally this will not be a shock for webui users.

I maybe should a add a quick note about the webui and a link to the webui howdoi

**Comment by elmiko** [comment]
> I maybe should a add a quick note about the webui and a link to the webui howdoi

++ adding another bullet point to the "additional resources" section would definitely be in order

**Comment by elmiko** [comment]
i'm good with the current changes, i think the language for the prerequisities could be tightened up a bit but i am good to merge this and adjust later.

**Comment by tmckayus** [comment]
@elmiko, tightened up how? we can fix it now

File: _howdoi/choose-my-spark-distribution.adoc [hunk 3]
```diff
@@ -0,0 +1,68 @@
+= How do I choose my own Apache Spark distribution?
+:page-layout: howdoi
+:page-menu_entry: How do I?
+
+When you install radanalytics.io, your project is set up to use a pre-determined distribution of Apache Spark. You can choose your own Apache Spark distribution with a few additional steps:
+
+.Prerequisites
+
+. Follow the instructions under link:/howdoi/install-radanalyticsio[How do I install radanalytics.io?]
```
Comment (elmiko) [inline _howdoi/choose-my-spark-distribution.adoc]: this should start with the prereq then lead into the instructions. similar to what we've done here https://radanalytics.io/howdoi/deploy-a-spark-cluster-cli

"An OpenShift project with the radanalytics.io manifest installed, see the Get Started instructions for more help."

File: _howdoi/choose-my-spark-distribution.adoc [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+. Choose an Apache Spark binary tarball like link:https://archive.apache.org/dist/spark/spark-2.3.0/spark-2.3.0-bin-hadoop2.7.tgz[this one]. You can download it or just use the url. Any binary distribution should work as long as it has this basic structure:
```
Comment (elmiko) [inline _howdoi/choose-my-spark-distribution.adoc]: same here, we should start with the prereq then lead into the discussion. so the prereq is "An Apache Spark tarball for the release you would like to deploy. Choose an Apache Spark tarball like...."

Comment (tmckayus) [reply _howdoi/choose-my-spark-distribution.adoc]: okay, gotcha, thanks

**Comment by tmckayus** [comment]
Answer to an earlier question: the simplest way to verify that your local imagestreams are being used is to use the oc client to dump yaml and compare between buildconfigs, pods, and imagestreams. Using the console to do this may be impossible or too cumbersome. For example

oc get is radanalytics-pyspark -o yaml
(under "status" this will list tags, the dockerImageReference field will tell you where the tagged image is from and list the sha256. The referenece should refer to the project. The sha256 is also given in the "image" field, incidentally)

oc get buildconfig sparkpi -o yaml
(under "triggers", imageChange.lastTriggeredImageID will show a reference that should match the imagestream dumped above)

oc get pod mycluster-w-1-p9lnw -o yaml
(under containerStatuses the imageID field will show a reference that should match a dump of the openshift-spark imagestream with "oc get is openshift-spark -o yaml")

A formal test can proabably be added based on this :)

**Comment by elmiko** [comment]
lGtm, thanks Trevor!

**Pull request merged**